File: sun4m/machine.py
from __future__ import annotations

import os
import struct

from .memory import SystemMemory, MMAP_BASE
from .elf import load_elf, load_elf_info, ElfInfo
from .cpu import CpuState


# Auxiliary vector types (from elf.h)
AT_NULL = 0  # End of vector
AT_IGNORE = 1  # Entry should be ignored
AT_EXECFD = 2  # File descriptor of program
AT_PHDR = 3  # Program headers for program
AT_PHENT = 4  # Size of program header entry
AT_PHNUM = 5  # Number of program headers
AT_PAGESZ = 6  # System page size
AT_BASE = 7  # Base address of interpreter
AT_FLAGS = 8  # Flags
AT_ENTRY = 9  # Entry point of program
AT_NOTELF = 10  # Program is not ELF
AT_UID = 11  # Real uid
AT_EUID = 12  # Effective uid
AT_GID = 13  # Real gid
AT_EGID = 14  # Effective gid
AT_PLATFORM = 15  # String identifying platform
AT_HWCAP = 16  # Machine-dependent hints about processor capabilities
AT_CLKTCK = 17  # Frequency of times()
AT_RANDOM = 25  # Address of 16 random bytes
AT_HWCAP2 = 26  # Extension of AT_HWCAP
AT_EXECFN = 31  # Filename of executable

# Interpreter base address (where we load ld-uClibc.so.0)
INTERP_BASE = MMAP_BASE  # 0x40000000


class Machine:

    def __init__(self, trace: bool = False, sysroot: str = ""):
        self.memory: SystemMemory = SystemMemory()
        # CpuState shares the same memory object to keep a single address space.
        self.cpu: CpuState = CpuState(memory=self.memory, trace=trace, sysroot=sysroot)
        self.entrypoint: int | None = None
        self.trace: bool = trace
        self.sysroot: str = sysroot

    def load_file(self, file: str, argv: list[str] | None = None) -> int:
        """Load an ELF file and set up execution environment.

        For dynamically linked executables, this also loads the interpreter
        and sets up the auxiliary vector for the dynamic linker.
        """
        with open(file, "rb") as f:
            elf_bytes = f.read()

        # Load the main executable
        main_info = load_elf_info(self.memory, elf_bytes, base_addr=0)
        self.entrypoint = main_info.entry_point

        # Load interpreter if present (dynamic linking)
        interp_info: ElfInfo | None = None
        if main_info.interpreter_path:
            interp_path = self._resolve_path(main_info.interpreter_path)
            if os.path.exists(interp_path):
                with open(interp_path, "rb") as f:
                    interp_bytes = f.read()
                interp_info = load_elf_info(
                    self.memory, interp_bytes, base_addr=INTERP_BASE
                )
                # For dynamic executables, start at interpreter entry point
                self.entrypoint = interp_info.entry_point

        # Initialise the CPU to start at the (possibly interpreter) entrypoint
        self.cpu.pc = self.entrypoint
        self.cpu.npc = (self.entrypoint + 4) & 0xFFFFFFFF

        # Initialize memory segments (after ELF is loaded to avoid conflicts)
        # Use a large stack for dynamic linking (16MB total)
        # Leave space above the "logical" stack top for SPARC ABI callers
        stack_size = 0x1000000  # 16MB
        stack_base = 0xD0000000
        self.cpu.memory.add_segment(stack_base, stack_size)
        # Set the logical stack top lower to leave headroom above SP
        stack_top = stack_base + stack_size - 0x100000  # Leave 1MB above

        # Add a segment at the top of address space for TLS/libc init
        self.cpu.memory.add_segment(0xFFFF0000, 0x10000)

        # Set up the stack with argc, argv, envp, auxv
        if argv is None:
            argv = [file]

        sp = self._setup_stack(
            stack_top, argv, main_info, interp_info
        )

        # Initialize %sp
        self.cpu.registers.write_register(14, sp)
        return self.entrypoint

    def _resolve_path(self, guest_path: str) -> str:
        """Resolve a guest path to host path using sysroot."""
        if self.sysroot and guest_path.startswith("/"):
            return self.sysroot + guest_path
        return guest_path

    def _setup_stack(
        self,
        stack_top: int,
        argv: list[str],
        main_info: ElfInfo,
        interp_info: ElfInfo | None,
    ) -> int:
        """Set up the stack with argc, argv, envp, auxv per SPARC ABI.

        Stack layout (growing down):
            [strings area - argv strings, env strings, platform string]
            [16 random bytes for AT_RANDOM]
            [padding for alignment]
            [auxv entries]
            [NULL - envp terminator]
            [envp pointers...]
            [NULL - argv terminator]
            [argv pointers...]
            [argc]
            [16-word register save area]  <- %sp points here

        Returns:
            The stack pointer value to use.
        """
        # Environment variables (for now, just empty or minimal)
        envp: list[str] = []

        # Reserve space at top of stack for various data
        # Start 256 bytes below stack_top for safety margin
        data_ptr = stack_top - 256

        # Write 16 random bytes for AT_RANDOM
        random_bytes = os.urandom(16)
        data_ptr -= 16
        data_ptr &= ~7  # Align to 8 bytes
        at_random_addr = data_ptr
        self.cpu.memory.write(at_random_addr, random_bytes)

        # Write platform string "sparc" for AT_PLATFORM
        platform = b"sparc\x00"
        data_ptr -= len(platform)
        platform_addr = data_ptr
        self.cpu.memory.write(platform_addr, platform)

        # Write executable filename for AT_EXECFN
        execfn = argv[0].encode("utf-8") + b"\x00"
        data_ptr -= len(execfn)
        execfn_addr = data_ptr
        self.cpu.memory.write(execfn_addr, execfn)

        # Write argv strings
        argv_string_addrs: list[int] = []
        for arg in argv:
            arg_bytes = arg.encode("utf-8") + b"\x00"
            data_ptr -= len(arg_bytes)
            self.cpu.memory.write(data_ptr, arg_bytes)
            argv_string_addrs.append(data_ptr)

        # Write envp strings
        envp_string_addrs: list[int] = []
        for env in envp:
            env_bytes = env.encode("utf-8") + b"\x00"
            data_ptr -= len(env_bytes)
            self.cpu.memory.write(data_ptr, env_bytes)
            envp_string_addrs.append(data_ptr)

        # Copy program headers to stack for AT_PHDR
        # The dynamic linker needs access to the main program's phdrs
        phdr_data = main_info.program_headers
        data_ptr -= len(phdr_data)
        data_ptr &= ~7  # Align
        phdr_stack_addr = data_ptr
        self.cpu.memory.write(phdr_stack_addr, phdr_data)

        # Align data_ptr for the arrays below
        data_ptr &= ~7

        # Build auxv entries
        auxv: list[tuple[int, int]] = []

        # AT_PHDR - address of program headers (on stack)
        auxv.append((AT_PHDR, phdr_stack_addr))
        # AT_PHENT - size of program header entry
        auxv.append((AT_PHENT, main_info.phdr_size))
        # AT_PHNUM - number of program headers
        auxv.append((AT_PHNUM, main_info.phdr_count))
        # AT_PAGESZ - page size
        auxv.append((AT_PAGESZ, 4096))
        # AT_BASE - interpreter base address (0 if no interpreter)
        auxv.append((AT_BASE, interp_info.base_address if interp_info else 0))
        # AT_FLAGS
        auxv.append((AT_FLAGS, 0))
        # AT_ENTRY - program entry point
        auxv.append((AT_ENTRY, main_info.entry_point))
        # AT_UID, AT_EUID, AT_GID, AT_EGID
        auxv.append((AT_UID, 1000))
        auxv.append((AT_EUID, 1000))
        auxv.append((AT_GID, 1000))
        auxv.append((AT_EGID, 1000))
        # AT_PLATFORM - platform string
        auxv.append((AT_PLATFORM, platform_addr))
        # AT_HWCAP - hardware capabilities (0 for basic SPARC)
        auxv.append((AT_HWCAP, 0))
        # AT_CLKTCK - clock ticks per second
        auxv.append((AT_CLKTCK, 100))
        # AT_RANDOM - 16 random bytes
        auxv.append((AT_RANDOM, at_random_addr))
        # AT_EXECFN - executable filename
        auxv.append((AT_EXECFN, execfn_addr))
        # AT_NULL - terminator
        auxv.append((AT_NULL, 0))

        # Calculate total size needed for pointers/values
        argc = len(argv)
        envc = len(envp)
        # Layout: 16 save words + argc + argv ptrs + NULL + envp ptrs + NULL + auxv
        # Each auxv entry is 2 words (type, value)
        pointer_area_size = (
            64  # 16-word register save area
            + 4  # argc
            + 4 * argc  # argv pointers
            + 4  # NULL
            + 4 * envc  # envp pointers
            + 4  # NULL
            + 8 * len(auxv)  # auxv entries (type + value each)
        )
        # Align to 8 bytes
        pointer_area_size = (pointer_area_size + 7) & ~7

        sp = data_ptr - pointer_area_size
        sp &= ~7  # Ensure 8-byte alignment

        # Write argc at sp + 64
        offset = 64
        self.cpu.memory.write(sp + offset, argc.to_bytes(4, "big"))
        offset += 4

        # Write argv pointers
        for addr in argv_string_addrs:
            self.cpu.memory.write(sp + offset, addr.to_bytes(4, "big"))
            offset += 4
        # NULL terminator for argv
        self.cpu.memory.write(sp + offset, (0).to_bytes(4, "big"))
        offset += 4

        # Write envp pointers
        for addr in envp_string_addrs:
            self.cpu.memory.write(sp + offset, addr.to_bytes(4, "big"))
            offset += 4
        # NULL terminator for envp
        self.cpu.memory.write(sp + offset, (0).to_bytes(4, "big"))
        offset += 4

        # Write auxv entries (pairs of 32-bit values)
        for atype, aval in auxv:
            self.cpu.memory.write(sp + offset, atype.to_bytes(4, "big"))
            offset += 4
            self.cpu.memory.write(sp + offset, (aval & 0xFFFFFFFF).to_bytes(4, "big"))
            offset += 4

        return sp

File: cpp/src/machine.cpp
#include "sun4m/machine.hpp"
#include "sun4m/endian.hpp"

#include <fstream>
#include <random>

namespace sun4m {

Machine::Machine(bool trace, std::string_view sysroot, std::vector<std::string> passthrough)
    : cpu(&memory, trace, sysroot, std::move(passthrough))
    , trace_(trace)
    , sysroot_(sysroot)
{
}

std::string Machine::resolve_path(std::string_view guest_path) const {
    if (!sysroot_.empty() && !guest_path.empty() && guest_path[0] == '/') {
        return sysroot_ + std::string(guest_path);
    }
    return std::string(guest_path);
}

std::expected<uint32_t, std::string>
Machine::load_file(const std::filesystem::path& file, const std::vector<std::string>& argv) {
    // Read ELF file
    std::ifstream ifs(file, std::ios::binary);
    if (!ifs) {
        return std::unexpected("Failed to open file: " + file.string());
    }
    std::vector<uint8_t> elf_bytes(
        (std::istreambuf_iterator<char>(ifs)),
        std::istreambuf_iterator<char>()
    );

    // Store the executable path for /proc/self/exe emulation
    cpu.exe_path = std::filesystem::absolute(file).string();

    // Load the main executable
    auto main_result = load_elf_info(memory, elf_bytes, 0);
    if (!main_result) {
        return std::unexpected(main_result.error());
    }
    auto& main_info = *main_result;
    entrypoint = main_info.entry_point;

    // Load interpreter if present (dynamic linking)
    std::optional<ElfInfo> interp_info;
    if (main_info.interpreter_path) {
        std::string interp_path = resolve_path(*main_info.interpreter_path);
        std::ifstream interp_ifs(interp_path, std::ios::binary);
        if (interp_ifs) {
            std::vector<uint8_t> interp_bytes(
                (std::istreambuf_iterator<char>(interp_ifs)),
                std::istreambuf_iterator<char>()
            );
            auto interp_result = load_elf_info(memory, interp_bytes, INTERP_BASE);
            if (interp_result) {
                interp_info = *interp_result;
                entrypoint = interp_info->entry_point;
            }
        }
    }

    // Initialize CPU to start at entry point
    cpu.pc = *entrypoint;
    cpu.npc = (*entrypoint + 4) & 0xFFFFFFFF;

    // Initialize memory segments
    // Stack: 16MB at 0xD0000000
    constexpr uint32_t stack_size = 0x1000000;
    constexpr uint32_t stack_base = 0xD0000000;
    memory.add_segment(stack_base, stack_size);
    uint32_t stack_top = stack_base + stack_size - 0x100000;  // Leave 1MB headroom

    // TLS segment at top of address space
    memory.add_segment(0xFFFF0000, 0x10000);

    // Set up stack with argc, argv, envp, auxv
    std::vector<std::string> actual_argv = argv.empty() ?
        std::vector<std::string>{file.string()} : argv;

    uint32_t sp = setup_stack(stack_top, actual_argv, main_info,
                              interp_info ? &*interp_info : nullptr);

    // Initialize %sp (register 14)
    cpu.registers.write_register(14, sp);

    // Initialize brk to after the loaded segments
    // Find the highest loaded address
    uint32_t max_addr = 0;
    for (const auto& [start, seg] : memory.segments()) {
        if (seg->end > max_addr && seg->start < MMAP_BASE) {
            max_addr = seg->end;
        }
    }
    cpu.brk = (max_addr + 0xFFF) & ~0xFFF;  // Page-align

    return *entrypoint;
}

uint32_t Machine::setup_stack(
    uint32_t stack_top,
    const std::vector<std::string>& argv,
    const ElfInfo& main_info,
    const ElfInfo* interp_info
) {
    // Empty environment for now
    std::vector<std::string> envp;

    // Reserve space at top of stack
    uint32_t data_ptr = stack_top - 256;

    // Write 16 random bytes for AT_RANDOM
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 255);
    std::array<uint8_t, 16> random_bytes;
    for (auto& b : random_bytes) {
        b = static_cast<uint8_t>(dis(gen));
    }
    data_ptr -= 16;
    data_ptr &= ~7;
    uint32_t at_random_addr = data_ptr;
    (void)memory.write(at_random_addr, random_bytes);

    // Write platform string "sparc"
    std::string platform = "sparc";
    data_ptr -= platform.size() + 1;
    uint32_t platform_addr = data_ptr;
    std::vector<uint8_t> platform_bytes(platform.begin(), platform.end());
    platform_bytes.push_back(0);
    (void)memory.write(platform_addr, platform_bytes);

    // Write executable filename for AT_EXECFN
    std::string execfn = argv.empty() ? "" : argv[0];
    data_ptr -= execfn.size() + 1;
    uint32_t execfn_addr = data_ptr;
    std::vector<uint8_t> execfn_bytes(execfn.begin(), execfn.end());
    execfn_bytes.push_back(0);
    (void)memory.write(execfn_addr, execfn_bytes);

    // Write argv strings
    std::vector<uint32_t> argv_string_addrs;
    for (const auto& arg : argv) {
        data_ptr -= arg.size() + 1;
        std::vector<uint8_t> arg_bytes(arg.begin(), arg.end());
        arg_bytes.push_back(0);
        (void)memory.write(data_ptr, arg_bytes);
        argv_string_addrs.push_back(data_ptr);
    }

    // Write envp strings
    std::vector<uint32_t> envp_string_addrs;
    for (const auto& env : envp) {
        data_ptr -= env.size() + 1;
        std::vector<uint8_t> env_bytes(env.begin(), env.end());
        env_bytes.push_back(0);
        (void)memory.write(data_ptr, env_bytes);
        envp_string_addrs.push_back(data_ptr);
    }

    // Align data_ptr
    data_ptr &= ~7;

    // Build auxv entries
    std::vector<std::pair<uint32_t, uint32_t>> auxv = {
        {AT_PHDR, main_info.phdr_addr},
        {AT_PHENT, main_info.phdr_size},
        {AT_PHNUM, main_info.phdr_count},
        {AT_PAGESZ, 4096},
        {AT_BASE, interp_info ? interp_info->base_address : 0},
        {AT_FLAGS, 0},
        {AT_ENTRY, main_info.entry_point},
        {AT_UID, 1000},
        {AT_EUID, 1000},
        {AT_GID, 1000},
        {AT_EGID, 1000},
        {AT_PLATFORM, platform_addr},
        {AT_HWCAP, 0},
        {AT_CLKTCK, 100},
        {AT_RANDOM, at_random_addr},
        {AT_EXECFN, execfn_addr},
        {AT_NULL, 0},
    };

    // Calculate total size needed
    size_t argc = argv.size();
    size_t envc = envp.size();
    size_t pointer_area_size = 64  // 16-word register save area
        + 4  // argc
        + 4 * argc  // argv pointers
        + 4  // NULL
        + 4 * envc  // envp pointers
        + 4  // NULL
        + 8 * auxv.size();  // auxv entries
    pointer_area_size = (pointer_area_size + 7) & ~7;

    uint32_t sp = data_ptr - static_cast<uint32_t>(pointer_area_size);
    sp &= ~7;

    // Write argc at sp + 64
    uint32_t offset = 64;
    std::array<uint8_t, 4> buf;
    write_be32(buf, static_cast<uint32_t>(argc));
    (void)memory.write(sp + offset, buf);
    offset += 4;

    // Write argv pointers
    for (uint32_t addr : argv_string_addrs) {
        write_be32(buf, addr);
        (void)memory.write(sp + offset, buf);
        offset += 4;
    }
    // NULL terminator for argv
    write_be32(buf, 0);
    (void)memory.write(sp + offset, buf);
    offset += 4;

    // Write envp pointers
    for (uint32_t addr : envp_string_addrs) {
        write_be32(buf, addr);
        (void)memory.write(sp + offset, buf);
        offset += 4;
    }
    // NULL terminator for envp
    write_be32(buf, 0);
    (void)memory.write(sp + offset, buf);
    offset += 4;

    // Write auxv entries
    for (const auto& [atype, aval] : auxv) {
        write_be32(buf, atype);
        (void)memory.write(sp + offset, buf);
        offset += 4;
        write_be32(buf, aval);
        (void)memory.write(sp + offset, buf);
        offset += 4;
    }

    return sp;
}

} // namespace sun4m
